File: swt/src/lcl/std.r/ap.f
      INTEGER MEM(10000)
      COMMON /DS$MEM/MEM
      INTEGER FD,ADDRE0,I,CODE,SIZE
      INTEGER OPEN,GETARG,GCTOI
      INTEGER ARG(102)
      CALL DSINIT(10000)
      IF((GETARG(1,ARG,102).NE.-1))GOTO 10000
        CALL ERROR('Usage: ap <file> [ <start address> ].')
10000 FD=OPEN(ARG,1)
      IF((FD.NE.-3))GOTO 10001
        CALL CANT(ARG)
10001 IF((GETARG(2,ARG,102).NE.-1))GOTO 10002
        ADDRE0=0
        GOTO 10003
10002   I=1
        ADDRE0=GCTOI(ARG,I,16)
10003 CALL LOAD(FD,CODE,SIZE)
      CALL RELOC0(FD,CODE,SIZE,ADDRE0)
      CALL DUMP(CODE,SIZE,ADDRE0)
      CALL CLOSE(FD)
      CALL SWT
      END
      SUBROUTINE LOAD(FD,CODE,SIZE)
      INTEGER FD,SIZE
      INTEGER CODE
      INTEGER MEM(10000)
      COMMON /DS$MEM/MEM
      INTEGER I,B
      INTEGER GETBY0
      INTEGER DSGET
      IF((GETBY0(B,FD).EQ.-1))GOTO 10005
      IF((B.NE.1))GOTO 10005
      GOTO 10004
10005   CALL ERROR('badly formed code file.')
10004 CALL GETWO0(SIZE,FD)
      CODE=DSGET(SIZE)
      I=0
      GOTO 10008
10006 I=I+1
10008 IF((I.GE.SIZE))GOTO 10007
        CALL GETBY0(MEM(CODE+I),FD)
      GOTO 10006
10007 RETURN
      END
      SUBROUTINE RELOC0(FD,CODE,SIZE,ADDRE0)
      INTEGER FD,SIZE,ADDRE0
      INTEGER CODE
      INTEGER MEM(10000)
      COMMON /DS$MEM/MEM
      INTEGER I,MAP,ADDR,UNNEC0
      INTEGER GETBY0
      IF((ADDRE0.NE.0))GOTO 10009
        RETURN
10009 IF((GETBY0(I,FD).EQ.-1))GOTO 10011
      IF((I.NE.2))GOTO 10011
      GOTO 10010
10011   CALL ERROR('badly formed code file.')
10010 CALL GETWO0(I,FD)
      I=0
      GOTO 10014
10012 I=I+1
10014 IF((I.GE.SIZE))GOTO 10013
        IF((MOD(I,8).NE.0))GOTO 10015
          CALL GETBY0(MAP,FD)
10015   IF((AND(MAP,LS(1,7-MOD(I,8))).EQ.0))GOTO 10016
          ADDR=ADDRE0+MEM(CODE+I+1)+LS(MEM(CODE+I),8)
          UNNEC0=CODE+I
          MEM(UNNEC0)=RS(ADDR,8)
          UNNEC0=UNNEC0+1
          MEM(UNNEC0)=RT(ADDR,8)
10016 GOTO 10012
10013 RETURN
      END
      SUBROUTINE DUMP(CODE,SIZE,ADDRE0)
      INTEGER CODE,SIZE,ADDRE0
      INTEGER MEM(10000)
      COMMON /DS$MEM/MEM
      INTEGER W
      INTEGER CRC
      INTEGER AAAAA0(20)
      INTEGER AAAAB0(10)
      INTEGER AAAAC0(11)
      DATA AAAAA0/170,180,172,173,177,182,172,176,233,160,170,180,172,17
     *7,182,172,176,233,160,0/
      DATA AAAAB0/170,178,172,173,177,182,172,176,233,0/
      DATA AAAAC0/160,170,180,172,173,177,182,172,176,233,0/
      CALL PRINT(-11,'L.')
      CALL PRINT(-11,AAAAA0,ADDRE0,SIZE)
      W=0
      GOTO 10019
10017 W=W+(1)
10019 IF((W.GE.SIZE))GOTO 10018
        CALL PRINT(-11,AAAAB0,MEM(W+CODE))
      GOTO 10017
10018 CALL PRINT(-11,AAAAC0,CRC(CODE,SIZE))
      RETURN
      END
      INTEGER FUNCTION CRC(CODE,SIZE)
      INTEGER CODE,SIZE
      INTEGER MEM(10000)
      COMMON /DS$MEM/MEM
      INTEGER MASK,SR,W,TEMP
      MASK=128
      SR=0
      W=0
10020   IF((MASK.NE.0))GOTO 10021
          W=W+(1)
          IF((W.LT.SIZE))GOTO 10022
            GOTO 10023
10022     MASK=128
10021   TEMP=0
        IF((AND(MEM(W+CODE),MASK).EQ.0))GOTO 10024
          TEMP=NOT(TEMP)
10024   IF((AND(SR,:100000).EQ.0))GOTO 10025
          TEMP=NOT(TEMP)
10025   IF((AND(SR,512).EQ.0))GOTO 10026
          TEMP=NOT(TEMP)
10026   IF((AND(SR,128).EQ.0))GOTO 10027
          TEMP=NOT(TEMP)
10027   IF((AND(SR,16).EQ.0))GOTO 10028
          TEMP=NOT(TEMP)
10028   IF((AND(SR,1).EQ.0))GOTO 10029
          TEMP=NOT(TEMP)
10029   SR=RS(SR,1)
        IF((TEMP.EQ.0))GOTO 10030
          SR=OR(SR,:100000)
10030   MASK=RS(MASK,1)
      GOTO 10020
10023 CRC=SR
      RETURN
      END
      SUBROUTINE GETBY0(B,FD)
      INTEGER B,FD
      INTEGER JUNK
      INTEGER MAPFD
      CALL PRWF$$(:1,MAPFD(FD),LOC(B),1,INTL(0),JUNK,JUNK)
      RETURN
      END
      SUBROUTINE GETWO0(W,FD)
      INTEGER W,FD
      INTEGER HI,LO
      CALL GETBY0(HI,FD)
      CALL GETBY0(LO,FD)
      W=OR(LS(HI,8),LO)
      RETURN
      END
C ---- Long Name Map ----
C address                        addre0
C getbyte                        getby0
C relocate                       reloc0
C getword                        getwo0
C unnecessary                    unnec0

File: swt/MFD/SRC/LCL/STD.R/AP.R
# ap --- generate object tape for A&P M6800 monitor

include PRIMOS_KEYS
include "as6800_def.r.i"

define(MEMSIZE,10000)

   DS_DECL (mem, MEMSIZE)

   integer fd, address, i, code, size
   integer open, getarg, gctoi

   character arg (MAXLINE)

   call dsinit (MEMSIZE)

   if (getarg (1, arg, MAXLINE) == EOF)
      call error ("Usage: ap <file> [ <start address> ].")

   fd = open (arg, READ)
   if (fd == ERR)
      call cant (arg)

   if (getarg (2, arg, MAXLINE) == EOF)
      address = 0
   else {
      i = 1
      address = gctoi (arg, i, 16)
      }

   call load (fd, code, size)
   call relocate (fd, code, size, address)
   call dump (code, size, address)

   call close (fd)
   stop
   end



# load --- load a code segment from file fd, return its locn and size

   subroutine load (fd, code, size)
   integer fd, size
   pointer code

   DS_DECL (mem, MEMSIZE)

   integer i, b
   integer getbyte

   pointer dsget

   if (getbyte (b, fd) == EOF || b ~= TEXTCODE)
      call error ("badly formed code file.")
   call getword (size, fd)

   code = dsget (size)
   for (i = 0; i < size; i = i + 1)
      call getbyte (mem (code + i), fd)

   return
   end



# relocate --- relocate code segment to specified address

   subroutine relocate (fd, code, size, address)
   integer fd, size, address
   pointer code

   DS_DECL (mem, MEMSIZE)

   integer i, map, addr, unnecessary
   integer getbyte

   if (address == 0)
      return

   if (getbyte (i, fd) == EOF || i ~= RMAPCODE)
      call error ("badly formed code file.")
   call getword (i, fd)

   for (i = 0; i < size; i = i + 1) {
      if (mod (i, 8) == 0)
         call getbyte (map, fd)
      if (and (map, ls (1, 7 - mod (i, 8))) ~= 0) {
         addr = address + mem (code + i + 1) + ls (mem (code + i), 8)
         unnecessary = code + i
         mem (unnecessary) = rs (addr, 8)
         unnecessary = unnecessary + 1
         mem (unnecessary) = rt (addr, 8)
         }
      }

   return
   end



# dump --- write relocated program in hex to STDOUT

   subroutine dump (code, size, address)
   integer code, size, address

   DS_DECL (mem, MEMSIZE)

   integer w
   integer crc

   call print (STDOUT, "L.")
   call print (STDOUT, "*4,-16,0i *4,16,0i "s, address, size)

   for (w = 0; w < size; w += 1)
      call print (STDOUT, "*2,-16,0i"s, mem (w + code))

   call print (STDOUT, " *4,-16,0i"s, crc (code, size))

   return
   end



# crc --- calculate Cyclic Redundancy Check polynomial

   integer function crc (code, size)
   integer code, size

   DS_DECL (mem, MEMSIZE)

   integer mask, sr, w, temp

   mask = 16r80
   sr = 0
   w = 0

   repeat {
      if (mask == 0) {
         w += 1
         if (w >= size)
            break
         mask = 16r80
         }
      temp = 0
      if (and (mem (w + code), mask) ~= 0)      temp = not (temp)
      if (and (sr, :100000) ~= 0)               temp = not (temp)
      if (and (sr, 16r0200) ~= 0)               temp = not (temp)
      if (and (sr, 16r0080) ~= 0)               temp = not (temp)
      if (and (sr, 16r0010) ~= 0)               temp = not (temp)
      if (and (sr, 16r0001) ~= 0)               temp = not (temp)
      sr = rs (sr, 1)
      if (temp ~= 0)
         sr |= :100000
      mask = rs (mask, 1)
      }

   crc = sr
   return
   end



# getbyte --- read one byte from code file

   subroutine getbyte (b, fd)
   integer b, fd

   integer junk
   integer mapfd

   call prwf$$ (KREAD, mapfd (fd), loc (b), 1, intl (0), junk, junk)

   return
   end



# getword --- read word from object code file

   subroutine getword (w, fd)
   integer w, fd

   integer hi, lo

   call getbyte (hi, fd)
   call getbyte (lo, fd)
   w = or (ls (hi, 8), lo)

   return
   end
